test/test_atacac/test_restore.py

```diff
@@ -0,0 +1,177 @@
+    gl = mock_gitlab.Gitlab.return_value
+    project = mock.MagicMock()
+    pipeline = mock.MagicMock()
+    job = mock.MagicMock()
+    job_content = mock.MagicMock()
+
+    gl.projects.get.return_value = project
+
+    project.pipelines.get.return_value = pipeline
+    project.jobs.get.return_value = job_content
+
+    pipeline.id = 12345
+    pipeline.attributes = {
+        "status": pipeline_status,
+        "sha": "value_sha",
+        "web_url": "value_web_url"
+    }
+    pipeline.jobs.list.return_value = [job]
+
+    job.name = artifact_job
+    job.attributes = {
+        "status": jop_status,
+        "web_url": "value_web_url"
+    }
+    job.artifacts = [
+        {
+            "filename": artifact_name,
+            "file_format": "value_file_format",
+            "size": "value_size"
+        }
+    ]
+
+    with tempfile.NamedTemporaryFile() as file:
+        try:
+            restore.gitlab_download(
+                "VALUE_url",
+                "VALUE_token",
+                "VALUE_project",
+                artifact_job,
+                "VALUE_pipeline_id",
+                file.name
+            )
+        except Abort:
+            assert (pipeline_status != "success"
+                    or jop_status != "success"
+                    or artifact_name != "artifacts.zip")
+
+
+@pytest.mark.parametrize(
+    "dir_exists",
+    [
+        pytest.param(True, id="destination is an existing directory"),
+        pytest.param(False, id="destination is an non-existing directory"),
+    ]
+)
+@mock.patch("atacac.restore.log")
+def test_extract(mock_log, dir_exists):
+    mock_dir = MockDir()
+    mock_dir.add_file(["A.yml", "X.yml", "B.yml", "Y.yml"])
+
+    dest_dir = mock_dir.path
+
+    shutil.make_archive("/tmp/atacac_test_archive", "zip", dest_dir)
+
+    if not dir_exists:
+        mock_dir.__del__()
+
+    restore.extract("/tmp/atacac_test_archive.zip", dest_dir)
+
+    assert sorted(os.listdir(dest_dir)) == sorted(["A.yml", "X.yml", "B.yml", "Y.yml"])
+    assert mock_log.call_count == 3 if dir_exists else 2
+
+
+@pytest.mark.parametrize(
+    "files",
+    [
+        pytest.param(
+            ["A.yml", "X.yml", "B.yml", "Y.yml"],
+            id="*.yml files"
+        ),
+        pytest.param(
+            ["A.yaml", "X.yaml", "B.yaml", "Y.yaml"],
+            id="*.yaml files (unsupported)"
+        ),
+        pytest.param(
+            ["A.xyz", "X.xyz", "B.xyz", "Y.xyz"],
+            id="*.xyz files (unsupported)"
+        ),
+    ]
+)
+@mock.patch("atacac.restore.log")
+@mock.patch("atacac.restore.tower_send")
+def test_upload(mock_tower_send, mock_log, files):
+    mock_dir = MockDir()
+    mock_dir.add_file(files)
+
+    restore.upload(mock_dir.path)
+
+    yml_files = [file for file in files if file.endswith(".yml")]
+
+    assert mock_log.call_count == len(yml_files) + 1
+    assert mock_tower_send.call_count == len(yml_files)
```

Comment:
You should check calls. You can't assume that `tower_send` was called with correct arguments.